sugar/lib/src/time/zone/providers/base_provider.dart

```diff
@@ -0,0 +1,45 @@
+import 'dart:collection';
+import 'package:sugar/src/time/zone/factory_timezone.dart';
+import 'package:sugar/src/time/zone/timezone.dart';
+
+/// Base class for all timezone providers.
+///
+/// This class is used to provide a Map interface, while using a cache
+/// to store the timezones that have already been fetched.
+/// This class also handles the 'Factory' timezone which is a special timezone
+/// which is treated separately from the other timezones.
+abstract class BaseTimezoneProvider extends UnmodifiableMapBase<String, Timezone> {
+  final _cache = <String, Timezone>{};
+
+  /// Returns a set of all known timezone identifiers.
+  Set<String> listTimezones();
+
+  /// Fetches the timezone with the given name.
+  Timezone? fetchTimezone(String name);
```

Comment:
Do we need this function if this is already a map?

Reply:
Again, this is needed so that all timezone providers include support for `Factory`

Reply:
Perhaps it might be simpler to add a line to the docs documenting the contract that all subclasses _must_ provide a factory timzone.

Reply:
IMHO, this is a better way to go about this.
Handling:
```
[](Object? key){
  if (key isI String){
    return ...
  }
}
```
in each subclass and reduplicating the `Factory` case for each subclass seems like a bad idea.

This is your project, if you want it removed, I'll do it.
Are you sure?

Reply:
I think the slight increase in duplication is worth the trade off of a reduced API surface.

Reply:
Done